pkg/allocator/cache.go

```diff
@@ -170,6 +183,8 @@ func (c *cache) onDeleteLocked(id idpool.ID, key AllocatorKey) {
 	delete(c.nextCache, id)
 	a.idPool.Insert(id)
 
+	c.emitChange(AllocatorChange{Kind: AllocatorChangeDelete, ID: id, Key: key})
```

Comment:
> This comment is actually unrelated from this line. A new key can be added into the key cache also through the `insert` function below, and a change will not be emitted in that case. Yet, when an observer is started and the current state is replayed, that event would be emitted, which is inconsistent.
> 
> Currently, an event is not sent to the events channel when `insert` is called, which might also need to be considered if planning to replace it with the observable pattern.

having a quick glance, it looks like `insert` gets called by `Allocate` which calls out to the backend via `lockedAllocate`. Eventually, after allocating the ID, the backend calls the cache via `OnAdd` (maybe `OnModify`) which triggers the observer. :thinking: But i didn't take a look at the branch with `localKeys`.

> Currently, an event is not sent to the events channel when insert is called, which might also need to be considered if planning to replace it with the observable pattern.

Above statement might also be the reason why no event is sent in case of an `insert`.

Reply:
I believe that above is correct--it's sort of like a "hairpin". Identity is created within one loop and once that creation event is received on the same node that created the identity, we then `emitChange`. I think the location of the latter is correct because we should essentially just mirror what `sendEvent` does.

Reply:
Thanks! My concern was mainly that, when replaying the state, an event would be emitted also for the entries that have been `inserted` but the corresponding notification has not yet been received from the kvstore. But at this point I guess it should not be a problem given that it will be eventually received (generating a spurious upsert event).